6x: Support receiving different Payload types

dsrees opened the pull request:
### What Changed
* Support receiving `message.payload` in a variety of types, support `Data`, Json `Any` or `Decodable` payload types
* Refactored how messages are deserialized coming in and serialized going out
* Defer payload deserialization to avoid round trip serialization

dsrees committed:
Copied over typed classes

dsrees committed:
Defer decoding payload as to not round trip serialization

dsrees committed:
Deferring decoding until binding callback

dsrees committed:
WIP: Trying to figure out why push isnt receiving replies

dsrees committed:
Pushes and joins

dsrees committed:
Deleting temp code

dsrees committed:
Moving to subscription and reusuing callbacks

dsrees committed:
Removed unused bindings from first pass

dsrees committed:
Replaced DecodedMessage with ReceivedMessage and MessageHeader

dsrees committed:
WIP OutgoingMessage

dsrees committed:
Renamed receivedMessage to incomingMessage

dsrees committed:
Further clean up of outoging messages

dsrees committed:
Removed JsonElement

dsrees committed:
Renamed TransportSerializer

dsrees committed:
Refactored PayloadDecoder

yeahphil commented:
Neat! I've been slammed with other stuff but took a quick look this afternoon.

I know it's WIP but I wonder if some things can now get collapsed. The two I'm thinking of are:
-  `Socket` can open many `Channel`s which can have many `ChannelSubscriptions` which can have many callbacks. Feels like there's an extra layer there. IMO a good way to solve this one would be to use `ChannelSubscription` internally, but keep public API on Channel rather than chaining calls.
- Parsing in `SubscriptionCallback` currently has a bunch of indirection (e.g. DataPayloadParser -> PhoenixPayloadDecoder -> JSONDecoder), but it could just use `JSONDecoder` directly, or `Channel` could pass down its decoder to allow configuration.

I think the overall direction is good, and a choose your own output approach to decoding seems like it's a good way to let legacy users upgrade but provide a path forward.

dsrees committed:
Cleaned up payload encoder/decoder interface

dsrees committed:
Completed tests for TransportSerializer

dsrees committed:
Cleaning up parsers

dsrees committed:
Parser tests

dsrees committed:
Moved decoders into trigger. Moved on calls in to extensions

dsrees committed:
PR Cleanup

dsrees commented:
@yeahphil some good feedback. Thanks

1. I've moved `ChannelSubscription` to be internal and kept the `channel.on("event") { ... }` syntax instead of the chaining syntax I proposed earlier
2. I've updated the encoder/decoder from socket to get passed down into the subscription callbacks to allow for configuration. 

Thanks again for your help with this stuff
